### PR TITLE
Story1task1

### PR DESCRIPTION
**Pull request by Landwaterfood**
- connected to database
- fetched the data
- make table for data with foreach
- seperate php.php and index. php now working
- seperate functions now
- tested

### CODE REVIEW

File: css.css
Comment (mallowmew) [inline css.css:1]: None of this CSS does anything on your page, why is it in your project?

File: index.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
+
+<!DOCTYPE html>
+<html lang="en">
+<head>
+
+    <?php require_once 'php.php'; ?>
```
Comment (mallowmew) [inline index.php]: PHP requires go at the top of the PHP file. These are not the same thing as HTML headers.

File: index.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            <th>Image Closeup</th>
+            <th>Image Site</th>
+            <th>Country</th>
+            <th>Town</th>
+            <th>Coordinates (Lat)</th>
+            <th>Coordinates (Long)</th>
+        </tr>
+        </thead>
+        <tbody>
+            <?php foreach ($collection as $pigment) :
```
Comment (mallowmew) [inline index.php]: Where does `$collection` come from? It is generally bad practice to `require` in variables, since people reading the code find it hard to follow. 

In this file you should call to a function that populates your `$collection` variable. This function can be `require`d and then used in this file.

**Review by mallowmew — CHANGES_REQUESTED** [review]
Not meeting definition of done:
* Not unit tested
* Not responsive

Database structure:
Not sure how country/town/coordinates relate to the topic. The name of the mineral/substance it's derived from and a closeup pic is probably enough.
If `color` is a colour category then it should be a linked / related table (categories are a one-to-many relationship).

**Comment by mallowmew** [comment]
It would be nice to see real data if this is a collection you are actually collecting. Make this app for yourself, you are the audience.

File: css.css
Comment (Landwaterfood) [reply css.css:1]: I just had it to hand to copy and paste... but perhaps ill just open it up, as and when, from the other project.

**Comment by Landwaterfood** [comment]
> Not meeting definition of done:
> 
> * Not unit tested
> * Not responsive
> 
> Database structure: Not sure how country/town/coordinates relate to the topic. The name of the mineral/substance it's derived from and a closeup pic is probably enough. If `color` is a colour category then it should be a linked / related table (categories are a one-to-many relationship).

Color is just a string with a color name in. It's just a way to have a broad category of "man colour names" (to use Matt's terminology). The HEX is a more precise/visual interpretation of the pigment.

**Comment by Landwaterfood** [comment]
> Not meeting definition of done:
> 
> * Not unit tested
> * Not responsive
> 
> Database structure: Not sure how country/town/coordinates relate to the topic. The name of the mineral/substance it's derived from and a closeup pic is probably enough. If `color` is a colour category then it should be a linked / related table (categories are a one-to-many relationship).

Im interested in the Geo Spatial aspects of where these pigments are found. A precise name for the collection app would be "Found pigments". Post FST, I want to learn QGIS and leaflet, the GIS stack to make interactive maps for pigment locations. This is a precursor to that :) But if you think I should just slim it down , keep it simple for this purpose, with the intention to expand the database fields for GeoSPatial, then that's fine.

**Comment by Landwaterfood** [comment]
> It would be nice to see real data if this is a collection you are actually collecting. Make this app for yourself, you are the audience.

I agree, but for I don't have the collection to hand and have vague data. Im looking forward to using the app whilst hunting of pigments.... I have a collection of books to hand, but that would be a bit lame. This idea seems exciting and generally quite useful for others too.

File: src/index.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+$collection= fetchPigmentData()
+
+
+?>
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <link rel="stylesheet" href="modern-normalize.css">
+    <link rel="stylesheet" href="stylesheet.css">
+    <meta name = "viewpoint" content = "width = device - width, initial-scale=1.0">
```
Comment (DBaldwin8) [inline src/index.php:11]: I was previously corrected that it should be "viewport" not "viewpoint"

Comment (DBaldwin8) [inline src/index.php]: I was previously corrected that it should be "viewport" not "viewpoint"

File: src/index.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            <th>Image Site</th>
+            <th>Country</th>
+            <th>Town</th>
+            <th>Coordinates (Lat)</th>
+            <th>Coordinates (Long)</th>
+        </tr>
+        </thead>
+        <tbody>
+        <?php echo createTable() ?>
+        </tbody>
```
Comment (DBaldwin8) [inline src/index.php]: I think your indentation needs looking at in the <thead> and <tbody> for easier reading.

File: src/php.php
Comment (DBaldwin8) [inline src/php.php:1]: good use of comments and ternary to check returns

File: tests/test.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        $expectedOutput .= '<td>Red</td>';
+        $expectedOutput .= '<td>#FF0000</td>';
+        $expectedOutput .= '<td>Iron Oxide</td>';
+        $expectedOutput .= '<td><img src="closeup.jpg" alt="Closeup view"></td>';
+        $expectedOutput .= '<td><img src="site.jpg" alt="Site view"></td>';
+        $expectedOutput .= '<td>USA</td>';
+        $expectedOutput .= '<td>Sedona</td>';
+        $expectedOutput .= '<td>34.8697</td>';
+        $expectedOutput .= '<td>-111.7608</td>';
+        $expectedOutput .= '</tr>';
```
Comment (DBaldwin8) [inline tests/test.php]: I think this could be refactored to not repeat $expectedOutput on every line (I can show you and example.)

**Comment by fennel3** [comment]
Could add types hints to functions! 
There is alot of location information going in, could this be condensed? such as coordinates lat and long? 
Unsure about the need for null values in the test!

Great work Christopher so far. Pigments will be documented!

**Review by DBaldwin8 — COMMENTED** [review]
added comments of changes I think. (hopefully this works, first time trying it)

File: src/index.php (hunk 3 above)
Comment (Landwaterfood) [reply src/index.php]: thanks for the tip!

File: src/php.php
Comment (Landwaterfood) [reply src/php.php:1]: thanks!

File: src/php.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        // Return an empty array if there's an error
+        return[];
+    }
+}
+function createpigmaDIVS(array $collection = null): string
+{
+    if (empty($collection) || !is_array($collection)) {
+
+
+        return '<div>Invalid pigment data.</div>';
```
Comment (DBaldwin8) [inline src/php.php:37]: remove white space above.

File: src/php.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        $result .=
+            '<div class = "pigment_item">' .
+            '<div class = "pigments pigment_id">' . '<div class = "stattitle">ID: #</div>' . $pigment['id'] . '</div>' .
+            '<div class = "pigments pigment_name">' . '<div class = "stattitle">name:</div>' . $pigment['name']  . '</div>' .
+            '<div class = "pigments pigment_color_name">' . '<div class = "stattitle">color: </div>' . $pigment['color_name'] . '</div>' .
+            '<div class = "pigments pigment_HEX">' . '<div class = "stattitle">HEX: </div>' . (is_null($pigment['HEX']) ? 'NULL' : htmlspecialchars($pigment['HEX'])) . '</div>' .
+            '<div class = "pigments pigment_mineral">' . '<div class = "stattitle">mineral: </div>' . (is_null($pigment['mineral']) ? 'NULL' : htmlspecialchars($pigment['mineral'])) . '</div>' .
+            '<div class = "pigments pigment_chemical">' . '<div class = "stattitle">chemical: </div>' . (is_null($pigment['chemical']) ? 'NULL' : htmlspecialchars($pigment['chemical'])) . '</div>' .
```
Comment (DBaldwin8) [inline src/php.php:68]: each concatenation on a new line for easier reading.

File: src/php.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    return $result;
+}
+
+
+
+
+
+
```
Comment (DBaldwin8) [inline src/php.php]: remove whitespace.

File: src/stylesheet.css [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        padding-right: 2%;
+        margin: 2%;
+    }
+
+
+    .stattitle{
+        display: inline-flex;
+        font-weight: bold;
+
+
```
Comment (DBaldwin8) [inline src/stylesheet.css:14]: inconsistent white space and indentation.

File: src/stylesheet.css [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+}
+.pigment_name{
+    font-weight: bold;
+}
+.pigment_color_name
+.pigment_HEX
+.pigment_mineral
+.pigment_chemical
+.pigment_description
+
```
Comment (DBaldwin8) [inline src/stylesheet.css]: If these aren't used shouldn't they be removed?

**Review by DBaldwin8 — COMMENTED** [review]
Suggestions below, all formatting suggestions.

File: src/php.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        echo 'connection error: ' . $exception->getMessage();
+        return null ;
+    }
+}
+
+function fetchPigmentData() : array
+{
+    // Establish a database connection
+    $connection = getConnected();
```
Comment (mallowmew) [inline src/php.php]: The connection to your database should go in the index.php, and then be passed in to this function.

This is so that when we have other functios that use the database, we don't make more connections - we pass the one we initially made through.

File: src/php.php [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    } catch (PDOException $exception)
+    {
+        echo 'connection error: ' . $exception->getMessage();
+        return null ;
+    }
+}
+
+function fetchPigmentData() : array
+{
+    // Establish a database connection
```
Comment (mallowmew) [inline src/php.php]: We generally avoid explanatory commentsbecause they are pretty redundant - we can see what the next line does by reading the function name. If the function name isn't clear enough to know what it does without a comment explaining it, consider changing the name of the function.

File: src/php.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        $query = $connection->prepare('SELECT pigment.id, pigment.name, pigment.HEX, pigment.mineral, pigment.chemical, pigment.description, pigment.image_closeup, color.name AS color_name
+            FROM pigment
+            INNER JOIN color color ON pigment.color_id = color.color_id');
+        $query->execute();
+        return $query->fetchALL();
+
+    } catch (PDOException $exception){
+        // In case of a database-related error, catch the exception and print an error message
+        echo 'fetch error:' . $exception->getMessage();
+        // Return an empty array if there's an error
```
Comment (mallowmew) [inline src/php.php]: Again clean up explanatory comments in production-ready code please - good to have in your notes, not necessary here.

File: src/php.php [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        return[];
+    }
+}
+function createpigmaDIVS(array $collection = null): string
+{
+    if (empty($collection) || !is_array($collection)) {
+
+
```
Comment (mallowmew) [inline src/php.php]: Remove unnecessary white space between if...return

File: src/php.php [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        return '<div>Invalid pigment data.</div>';
+    }
+    $result = "";
+
+    foreach ($collection as $pigment) {
+        if(!isset($pigment['id']) || is_null($pigment['id'])){
+            return '';
+        }
+        if(!is_int($pigment['id'])){
+            throw new UnexpectedValueException('id was not a num');
```
Comment (mallowmew) [inline src/php.php]:
```suggestion
            throw new UnexpectedValueException('id was not a number');
```
We generally avoid shortening words because it quickly gets confusing when you have to read this error message. The infamous example is when people shorten 'category' to 'cat' - short for category? something to do with animals? an acronym? It's unnecessary to over-shorten your words.

File: src/stylesheet.css (hunk 9 above)
Comment (mallowmew) [reply src/stylesheet.css:14]: I agree with Darren, choose a consistent line-breaking and indentation style please.

File: src/stylesheet.css (hunk 10 above)
Comment (mallowmew) [reply src/stylesheet.css]: Yes they should, no empty CSS groups in CSS files.

File: src/php.php [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        return $query->fetchALL();
+
+    } catch (PDOException $exception){
+        // In case of a database-related error, catch the exception and print an error message
+        echo 'fetch error:' . $exception->getMessage();
+        // Return an empty array if there's an error
+        return[];
+    }
+}
+function createpigmaDIVS(array $collection = null): string
```
Comment (mallowmew) [inline src/php.php:34]: I'm not a big fan of this function name for a few reasons:
* 'pigma' seems like a shortened word
* 'pigma' is not snake_cased OR camelCased to distinguish it as a word
* DIVS is in ALLCAPS but it isn't an acronym
* 'create pgima divs' is close to being a technical detail - `displayPigments` would be a more conventional name that tells you *what the function does* and not *how it does it* - in programming, the *how* things are done often changes as an application is developed, but the 'what' rarely changes.

File: src/php.php (hunk 8 above)
Comment (mallowmew) [reply src/php.php]: Seconded,

File: tests/test.php
Comment (mallowmew) [inline tests/test.php:1]: PHPUnit cannot find your tests when you run `phpunit tests/` because there are some rules about file naming you are not following:

* Files with `class`es in them must be named the same as the class
* PHPUnit test filenames must end with the suffix `Test`

Please rename this file to `createpigmaDIVSTest.php` and then PHPUnit will be able to pick it up properly.

File: tests/test.php [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+            ];
+
+
+        $expectedOutput = '';
+
+        $actualOutput = createpigmaDIVS($mockData);
+        $this->assertEquals($expectedOutput, $actualOutput);
+    }
+
+    public function testcreatepigmaDIVSrightData(): void
```
Comment (mallowmew) [inline tests/test.php]: What kind of test is this? Is it a success, a failure, or a malformed test?

It would be standard to name the test function like 
```suggestion
    public function testcreatepigmaDIVS_success(): void
```

So that the kind of test is clearly included

File: tests/test.php [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+            ]
+        ];
+
+        $expectedOutput ='<div class = "pigment_item"><div class = "pigments pigment_id"><div class = "stattitle">ID: #</div>1</div><div class = "pigments pigment_name"><div class = "stattitle">name:</div>Red Ochre</div><div class = "pigments pigment_color_name"><div class = "stattitle">color: </div>Red</div><div class = "pigments pigment_HEX"><div class = "stattitle">HEX: </div>#FF0000</div><div class = "pigments pigment_mineral"><div class = "stattitle">mineral: </div>NULL</div><div class = "pigments pigment_chemical"><div class = "stattitle">chemical: </div>FE02</div><div class = "pigment_description"><div class = "stattitle">description: </div>pigment from the ground</div><div class = "pigment_image_closeup"><img src="closeup.jpg" alt="image" style="width:100%;height:80%;"></div></div>';
+
+        $actual = createpigmaDIVS($mockData);
+        $this->assertEquals($expectedOutput, $actual);
+    }
+
+    public function testcreatepigmaDIVSWithNullValues(): void
```
Comment (mallowmew) [inline tests/test.php]: Is it a success, failure or malformed test?

File: tests/test.php [hunk 19]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+use PHPUnit\Framework\TestCase;
+
+require_once 'src/php.php';
+
+class createpigmaDIVStest extends TestCase
+{
+    public function testcreatepigmaDIVSWrongdata(): void
```
Comment (mallowmew) [inline tests/test.php]: Is it a success, failure or malformed test?

**Review by mallowmew — CHANGES_REQUESTED** [review]
